### PR TITLE
WorkflowPluginTest.sshGitInsideDocker overhaul

### PR DESCRIPTION
**Pull request by jglick**
@mafraba found that this test could fail in a Dockerized test environment and tried to work around this by running the test environment with `-v /tmp`, without success. At root the problem was the unrealistic structure of the test, in which the build was being run on master, using the test environment’s own filesystem; and the consequent attempt to work around JENKINS-36997 by using `/tmp` for the workspace, which will not generally be available in a sibling container. To make it all work I had to switch to running the build in an agent container (linked to the Git container) accessing the host Docker daemon, a supported (indeed recommended) mode for `docker-workflow`, which does not trigger JENKINS-36997 since the workspace path is guaranteed to be short.

@reviewbybees

### CODE REVIEW

File: src/test/java/plugins/WorkflowPluginTest.java [hunk 1]
```diff
@@ -196,30 +201,49 @@
...
+        SshSlaveLauncher launcher = slave.setLauncher(SshSlaveLauncher.class);
+        Process proc = new ProcessBuilder("stat", "-c", "%g", "/var/run/docker.sock").start();
+        String group = IOUtils.toString(proc.getInputStream()).trim();
+        Assume.assumeThat("docker.sock can be statted", proc.waitFor(), is(0));
+        try {
+            Integer.parseInt(group);
+        } catch (NumberFormatException x) {
+            Assume.assumeNoException("unexpected output from stat on docker.sock", x);
+        }
+        // Note that we need to link to the Git container both on the agent, for the benefit of JGit (TODO pending JENKINS-30600), and in the build container, for the benefit of git-pull:
```
Comment (jglick) [inline src/test/java/plugins/WorkflowPluginTest.java]: If you disable the JENKINS-30600 workarounds, you get the expected

```
Started by user anonymous
[Pipeline] node
Running on … in /home/test/workspace/…
[Pipeline] {
[Pipeline] sh
[…] Running shell script
+ docker inspect -f . cloudbees/java-build-tools
.
[Pipeline] withDockerContainer
$ docker run -t -d -u … --link=…:git -w /home/test/workspace/… --volumes-from … -e … --entrypoint cat cloudbees/java-build-tools
[Pipeline] {
[Pipeline] git
Cloning the remote Git repository
Cloning repository ssh://git@git/home/git/gitRepo
 > git init /home/test/workspace/… # timeout=10
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Could not init /home/test/workspace/…
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$5.execute(CliGitAPIImpl.java:663)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:464)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
	at hudson.remoting.UserRequest.perform(UserRequest.java:153)
	at …
	at ......remote call to …(Native Method)
	at …
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
	at …
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
	at com.sun.proxy.$Proxy79.execute(Unknown Source)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1057)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1097)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:109)
	at …
Caused by: hudson.plugins.git.GitException: Error performing command: git init /home/test/workspace/…
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1791)
	at …
Caused by: java.io.IOException: Cannot run program "git" (in directory "/home/test/workspace/…"): error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at hudson.Proc$LocalProc.<init>(Proc.java:245)
	at hudson.Proc$LocalProc.<init>(Proc.java:214)
	at hudson.Launcher$LocalLauncher.launch(Launcher.java:846)
	at hudson.Launcher$ProcStarter.start(Launcher.java:384)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1780)
	... 15 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 20 more
[Pipeline] }
$ docker stop --time=1 …
$ docker rm -f …
[Pipeline] // withDockerContainer
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: null
Finished: FAILURE
```

and thus might serve as an integration test for https://github.com/jenkinsci/git-client-plugin/pull/219 (CC @ndeloof).

File: src/test/java/plugins/WorkflowPluginTest.java [hunk 2]
```diff
@@ -196,30 +201,49 @@
...
+        GitContainer gitContainer = gitServer.get();
         GitRepo repo = new GitRepo();
         repo.commit("Initial commit");
-        repo.transferToDockerContainer(host, port);
+        repo.transferToDockerContainer(gitContainer.host(), gitContainer.port());
+        DumbSlave slave = jenkins.slaves.create(DumbSlave.class);
+        slave.setExecutors(1);
+        slave.remoteFS.set("/home/test"); // TODO perhaps should be a constant in SshdContainer
+        SshSlaveLauncher launcher = slave.setLauncher(SshSlaveLauncher.class);
+        Process proc = new ProcessBuilder("stat", "-c", "%g", "/var/run/docker.sock").start();
```
Comment (jglick) [inline src/test/java/plugins/WorkflowPluginTest.java]: Works for me on Linux; might cause the test to be skipped on some environments (e.g., boot2docker), but I cannot test these.

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/DockerAgentContainer/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+groupadd -g $DOCKER_GROUP docker
+usermod -aG docker test
+chown test.test /home/test/workspace
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/DockerAgentContainer/entrypoint.sh]: cf. `VOLUME`

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/DockerAgentContainer/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,9 @@
+FROM jenkins/java:eed7706a10fb
+
+RUN cd /tmp && \
+    wget -nv -O - https://get.docker.com/builds/Linux/x86_64/docker-1.13.1.tgz | tar xvfz - docker/docker && \
+    chmod a+x docker/docker && \
+    mv docker/docker /usr/bin/docker
+VOLUME /home/test/workspace
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/DockerAgentContainer/Dockerfile]: so we have something to share with the build’s image

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/DockerAgentContainer/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,4 @@
+groupadd -g $DOCKER_GROUP docker
+usermod -aG docker test
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/DockerAgentContainer/entrypoint.sh]: cf. `stat` during test (`docker.sock` should not generally be writable by other users)

File: src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/GitContainer.java [hunk 6]
```diff
@@ -32,9 +32,17 @@ public String getRepoUrl() {
         return "ssh://git@" + host() + ":" + port() + REPO_DIR;
     }
 
-    /** URL visible from other Docker containers. */
+    @Deprecated
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/GitContainer.java]: not reliable without `--link`

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainerHolder.java [hunk 7]
```diff
@@ -45,23 +45,31 @@
...
+                throw new Error("Failed to start container - " + type, e);
             }
         }
         return container;
     }
 
     /**
+     * Provides a starter directly, so you can customize it a bit before calling {@link Starter#start}.
+     */
+    public Starter<T> starter() throws IOException, InterruptedException {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainerHolder.java]: extracting for use from new test; behavior for tests using `.get()` should be unchanged

File: src/test/java/plugins/WorkflowPluginTest.java [hunk 8]
```diff
@@ -196,30 +201,49 @@
 
     @Category(DockerTest.class)
     @WithDocker
-    @WithPlugins({"workflow-aggregator@1.14", "docker-workflow@1.4", "git", "ssh-agent@1.10"})
+    @WithPlugins({"workflow-cps@2.0", "workflow-job@2.0", "workflow-durable-task-step@2.0", "docker-workflow@1.5", "git", "ssh-agent@1.10", "ssh-slaves@1.11"})
```
Comment (jglick) [inline src/test/java/plugins/WorkflowPluginTest.java]: `workflow-aggregator` pulls in many deps we do not need here, slowing down test initialization

**Comment by jglick** [comment]
@reviewbybees done

**Comment by recena** [comment]
@jglick Thanks so much!

**Pull request merged**